crates/runtime/src/traphandlers/macos_aarch64.S

```diff
@@ -0,0 +1,71 @@
+.section __TEXT,__text,regular,pure_instructions
+
+#define GLOBL(fnname) .globl _##fnname
+#define HIDDEN(fnname) .private_extern _##fnname
+#define TYPE(fnname)
+#define FUNCTION(fnname) _##fnname
+#define SIZE(fnname)
+
+HIDDEN(wasmtime_macos_aarch64_unwind)
+GLOBL(wasmtime_macos_aarch64_unwind)
+
+// This small stub helps connect the native code and the wasm call frames when
+// unwinding a stack. This must be kept in close sync with macos.rs in the same
+// directory.
+
+// We expect the following values in the associated argument registers:
+// wasmtime_macos_aarch64_unwind(faulting_pc, setjmp_buf, unwind_fn)
+
+.p2align 2
+TYPE(wasmtime_macos_aarch64_unwind)
+FUNCTION(wasmtime_macos_aarch64_unwind):
+.cfi_startproc simple
+    // Set the pseudo-register RA_SIGN_STATE to 0. This disables pointer
+    // authentication of the return address for this particular stub.
+    .cfi_escape 0x16,    /* DW_CFA_val_expression */ \
+        0x22,            /* pseudo-reg 34 is RA_SIGN_STATE */ \
+        1,               /* the byte length of this expression */ \
+        0x30             /* DW_OP_lit0 */
+
+    // This directive does the magic. In general, when unwinding aarch64 call
+    // frames, LR == PC. But when unwinding wasm frames, the PC may be at a
+    // value that's strictly different from LR. So first we want to unwind the
+    // frame designated by PC, and then only use LR to unwind to its caller.
+    //
+    // This directive says that the return register for this particular frame
+    // is located in x3. The choice of x3 is arbitrary among the possible
+    // available registers; x0, x1 and x2 can't be used, because they hold
+    // function arguments.
+    //
+    // Then, the cfi_offset directive below defines where x3 is located: this
+    // matches the stack slot where we've put PC, in macos.rs, so we're really
+    // indicating that the next call frame information must be found using this
+    // PC.
+    //
+    // Other things that have been tried and did not work:
+    // - not allocate a stack frame, and have one register contain the value of
+    // PC (neither cfi_register or cfi_same_value worked).
+    // - setting LR to the faulting PC: not sufficient, because the blr below
+    // will clobber it, resulting in the faulting call frame being lost when
+    // unwinding.
+    .cfi_return_column 3
+
+    // Define what's the CFA == the stack pointer of whatever called into here.
+    // Here, it's defined as ($sp + 0x10), as we've allocated 16 bytes in the
+    // resume function.
+    .cfi_def_cfa x31, 0x10
+
+    // PC has been saved at CFA - 0x10.
+    .cfi_offset x3, -0x10
+
+    // The unwind function is stored in x2. Other arguments to this function
+    // are already in the right places (x0 and x1). We can branch and link to
+    // x2 (LR has been explicitly saved on the stack by Cranelift).
+    blr x2
```

Comment:
This feels somewhat iffy to me that we rely on Cranelift to restore `lr` to unwind since this `blr` is otherwise overwriting the value of the wasm frame's `lr`. In the future if we optimize cranelift to not save `lr` to the stack for leaf functions, for example, we'd have to update this?

I figured it was easy enough to push `lr` to the stack and have a CFI directive for how to restore it?

Reply:
As discussed on zulip, turns out we can even:

- set `lr` from the faulting pc
- jump to unwind directly
- burn the assembly stub with 🔥 

I tried doing something like this a week or so ago, but it wasn't working (probably because of the intertwined return-address signing issue!). Good riddance!